### ISSUE TITLE
proposal: stop using singletons, store config as tree of configs

### ISSUE THREAD

**Issue by abelnation** [bcoe/yargs#347]
So my goal was to use yargs (whose API I really like) to build a repl with auto-complete functionality.  There were a number of problems I observed in the yargs code that made this hard.

* The yargs parser is a global singleton, and uses `reset()` to move between different config states
* Parsing a subcommand calls `reset()` on yargs, blowing away your previous configs.  In a repl context, one potentially would be parsing multiple lines over the course of a single program.  
* I wanted the completion api to be really easy to call, so I could use it in my repl
* Contextual auto-completion for commands was clunky in the current version since it doesn't take into account the full context (i.e. only complete valid commands given the previous commands in the line)

I put together a new implementation of the API that tries to address these problems.  I implement the main elements of the API, but have left out some of the more extraneous calls for now.  The rest of the api would be trivial to implement.

https://github.com/abelnation/yargs/tree/migrate-to-es6

Some notes about my implementation:
- It is written in es-2015 (es6) javascript, and uses babel to compile down to vanilla JS.  I find that es6 js allows for much more concise and readable code, especially with respect to de-structuring arguments.  I kept `standard` in place, so stylistically, the code should be consistent.
- see `./src/index.es6` as the starting point, run `npm run compile` to compile es6 to vanilla js
- relevant npm commands have been added for compiling and linting
- I added an `ArgConsumer` utility class (basically, a specialized iterator) which has really helped with consuming different amounts of tokens at a time from the line (e.g. options with an argument / multiple arguments / none, etc.)
- It is set up to write mocha tests in either vanilla or es6 js, I just haven't written any real tests yet.  

Would love to hear your thoughts.  I think the usage and validation stuff is really strong, and would love to help contribute to the effort.

**Comment by bcoe** [comment]
@abelnation this is awesome, but definitely more of a leap forward than I would entertain in one swoop f

**Comment by abelnation** [comment]
haha, yeah, it started as a fun project to just write my own, but i think it's kind of pointless to *totally* re-invent the wheel :P

i do think the whole singleton thing is a big problem and makes the API needlessly complicated for more than just basic CLI usage.  

i think my implementation actually simplifies some aspects of the parsing and completions quite a bit (once you get past the es6 migration).  Already, the number of code lines is reduced significantly.  I'd def encourage you to take a good look, and I'd be down to put in effort to integrate in a more incremental way  (as opposed to one fell swoop).

**Comment by abelnation** [comment]
Another enhancement I added to the parsing: when your argv has commands, the parse result contains a  key, `command`, which has the commands and subcommands concatted together for easy grabbing, e.g. `foo bar --force 1 2 3` where `foo` and `bar` are commands would yield:

```
{
  _: [ 1, 2, 3 ],
  force: true,
  command: 'foo bar'
}
```

**Comment by bcoe** [comment]
@abelnation I was on my phone and hit the wrong button, I will provide better feedback by end of day.

**Comment by abelnation** [comment]
haha for sure

**Comment by bcoe** [comment]
@abelnation sorry that it's taken me a bit longer to get your feedback than I claimed it would:

* I love the idea of breaking yargs of its singleton behavior -- I wonder if there's a way we can do this without breaking the existing contract. For instance, `index.js` could continue to be a singleton, but it would import a class that can be instantiated.
  * the idea of using yargs inside a REPL environment is really neat, and I'd love to figure out how we might facilitate this.
* I've played a bit with ES2015/babel (I actually added support for this to the [nyc](https://github.com/bcoe/nyc) test coverage tool I maintain). I think some of the constructs that it introduces are really cool, and I'm excited to continue to familiarize myself with ES2015. However, I'm not quite prepared to add a build step to the yargs project.
  * if you're running web-services on `noce@4+` it seems quite reasonable to start using the ES2015 constructs that are in the bleeding edge `v8`. For a glue library like yargs that runs on every version of Node.js, however, I'm not too excited to add `babel`. This is a personal preference, not a value judgement :smile:
* we actually already use the `standard` coding style for yargs, I'm a big fan of this:
  * it's a coding style that's getting traction with a lot of other repos, e.g., npm itself.
  * it removes any arguments about code style from this repo, folks can go argue until the cows come home on the `standard` repo :+1: 

In summary:

* I love the idea of solving the singleton problems that make yargs hard to use in a REPL -- I think solving some of these problems would help us start to improve command functionality (a feature that we need to do some hard thinking about).
* I would like to patch any bugs with completion that you've run into (and it would be great to abstract completion as you suggest in https://github.com/bcoe/yargs/issues/335)
* I'm not too keen to move to ES2015, but let's revisit this in a year.

I would love your help on the `v4` branch adding test cases around your requirements for completion and moving yargs away from being a singleton, but let's move one small pull at a time. And let's keep a lively conversation going here:

https://github.com/bcoe/yargs/issues/310

**Comment by bcoe** [comment]
closin' but with notes; please feel free to open other tickets, and to help build some of these features.
